### PR TITLE
Move TransformCorrelatedScalarAggregationToJoin to the iterative optimizer

### PR DESCRIPTION
**Pull request by maciejgrzybek**
Fix from https://github.com/maciejgrzybek/presto/commit/77deae9ff0384b9c0c7f9fb76a7e5b73e8df5f84 is nothing specific to the new optimizer. Previous optimizer (in the old framework) was also not preserving the invariant of output symbols but we didn't have a check for that so we never noticed.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java [hunk 1]
```diff
@@ -165,10 +168,25 @@ private PlanNode rewriteScalarAggregation(
...
+                    idAllocator.getNextId(),
+                    aggregationNode.get(),
+                    assignments);
         }
     }
 
+    private static List<Symbol> getTruncatedAggregationSymbols(ApplyNode applyNode, AggregationNode aggregationNode)
+    {
+        Set<Symbol> applySymbols = new HashSet<>(applyNode.getOutputSymbols());
+        return aggregationNode.getOutputSymbols().stream()
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java]: What if `AggregationNode` does not produce some symbol from `ApplyNode`'s output symbols (e.g. due to an error in rewrite)? With your approach, optimizer's sanity check wouldn't catch that

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java [hunk 2]
```diff
@@ -44,6 +44,7 @@
 import com.google.common.collect.ImmutableMap;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java]: Good catch

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarAggregationToJoin.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        if (applyNode.getCorrelation().isEmpty() || !(isScalar(subquery) && applyNode.isSubqueryResolved())) {
+            return Optional.empty();
+        }
+
+        Optional<AggregationNode> aggregation = findAggregation(subquery, lookup);
+        if (!(aggregation.isPresent() && aggregation.get().getGroupingKeys().isEmpty())) {
+            return Optional.empty();
+        }
+
+        ScalarAggregationToJoinRewriter rewriter = new ScalarAggregationToJoinRewriter(functionRegistry, symbolAllocator, idAllocator, lookup);
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarAggregationToJoin.java]: Maybe we should put a not, that this class must be moved here after we get rid of the old optimizer?

Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarAggregationToJoin.java]: Done. Added a note here: https://github.com/prestodb/presto/pull/7429/files#diff-47cda7391e86cd5670b10e551ec55311R63

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java [hunk 4]
```diff
@@ -179,7 +206,7 @@ private PlanNode replaceFirstRecursive(PlanNode node, PlanNode nodeToReplace)
             return node;
         }
         else if (sources.size() == 1) {
-            return replaceChildren(node, ImmutableList.of(replaceFirstRecursive(node, sources.get(0))));
+             return replaceChildren(node, ImmutableList.of(replaceFirstRecursive(node, sources.get(0))));
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: No. Thanks, fixed!

**Comment by maciejgrzybek** [comment]
Resolved conflicts and rebased on master.

Did you have a chance to have a look on that, @martint or @kokosing?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java [hunk 5]
```diff
@@ -29,18 +30,32 @@
...
+
     private final PlanNode node;
+    private final Lookup lookup;
     private Predicate<PlanNode> where = alwaysTrue();
     private Predicate<PlanNode> skipOnly = alwaysTrue();
 
+    @Deprecated
     public PlanNodeSearcher(PlanNode node)
     {
+        this(node, (n) -> n);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: use `Function::identity`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java [hunk 6]
```diff
@@ -29,18 +30,32 @@
 
 public class PlanNodeSearcher
 {
+    @Deprecated
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: I am not sure if it is the way to go to use `PlanNodeSearcher` in the iterative optimizer. I think this rule depends on having pattern matching and traits to be implemented first.

@martint What do you think?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java (hunk 5 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: I can't. It's `Lookup` interface. Lambda here is just a shortcut to anonymous class implementation.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java (hunk 6 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: @kokosing I think of traits as an extension to the iterative optimizer. We should be aiming to enable iterative optimizer as the default one, then we can refactor/revamp it's traversal strategy with traits etc. The sooner we enable iterative optimizer as default the sooner it becomes robust.
Therefore, stalling the process of porting of some rules to the moment when we have traits, is worse strategy then porting them and refactoring later, when new features to the optimizer are in. I think baby steps are always better.

**Comment by maciejgrzybek** [comment]
@kokosing, @martint: thoughts?

**Comment by maciejgrzybek** [comment]
Rebased on master to resolve conflicts.
Any further thoughts, guys? @martint, @kokosing?

**Comment by maciejgrzybek** [comment]
Rebased onto master to resolve conflicts (only imports were conflicting, though).

**Comment by maciejgrzybek** [comment]
Any thoughts?
Please let me know when you want to review it, so I won't be rebasing it unnecessarily.

**Comment by martint** [comment]
I have a few PRs in my queue before this one. I'll let you know when I start reviewing it.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java [hunk 7]
```diff
@@ -36,8 +37,13 @@ private DependencyExtractor() {}
 
     public static Set<Symbol> extractUnique(PlanNode node)
     {
+        return extractUnique(node, n -> n);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java]: `n -> n` is not sufficient Lookup stub here, as in `ExpressionExtractor`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionExtractor.java [hunk 8]
```diff
@@ -25,13 +27,22 @@
...
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
 
 public class ExpressionExtractor
 {
     public static List<Expression> extractExpressions(PlanNode plan)
     {
+        return extractExpressions(plan, n -> n);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionExtractor.java]: `n -> n` is not sufficient Lookup stub here. With it, `ExpressionExtractor.Visitor#visitPlan` recurses till the end of the stack when it encounters `GroupReference`. https://github.com/prestodb/presto/pull/7982/commits/77ec5e9c5cacc7d41811be6f133b1b117d21fd7b would be more appropriate.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java [hunk 9]
```diff
@@ -36,8 +37,13 @@ private DependencyExtractor() {}
...
+    }
+
+    public static Set<Symbol> extractUnique(PlanNode node, Lookup lookup)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java]: Before this change, a `Rule` was able to extract symbols from a `PlanNode` without recursing into sources. This property turned out to be important (https://github.com/prestodb/presto/pull/7888/files#diff-1a430cab232220a3d693ed1d27066e9fR377), so could you retain `extractUnique` overload that does that?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+                if (node.getType() != LogicalBinaryExpression.Type.AND) {
+                    context.set(false);
+                }
+                return null;
+            }
+        }.process(predicate, isSupported);
+        return isSupported.get();
+    }
+
+    private static class DecorrelatedNode
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java]: https://github.com/prestodb/presto/pull/7888/commits/690894b11641c445aebc41c4a389bf83c5e31a08#diff-1a430cab232220a3d693ed1d27066e9fR404 ... Do you see potential for re-use?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 11]
```diff
@@ -161,7 +161,7 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                 new RemoveUnreferencedScalarInputApplyNodes(),
                 new TransformUncorrelatedInPredicateSubqueryToSemiJoin(),
                 new TransformUncorrelatedScalarToJoin(),
-                new TransformCorrelatedScalarAggregationToJoin(metadata),
+                new TransformCorrelatedScalarAggregationToJoin(metadata.getFunctionRegistry()),
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Is there gain from passing FR instead of Metadata? Can someone in the future feel need to undo this change because code will need more info?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java [hunk 12]
```diff
@@ -23,23 +25,44 @@
...
+        {
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
         @Override
         protected Boolean visitPlan(PlanNode node, Void context)
         {
+            if (node instanceof GroupReference) {
+                checkState(lookup.isPresent(), "Lookup has to be provided to test plans with GroupReference nodes");
+                return lookup.get().resolve(node).accept(this, null);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: https://github.com/prestodb/presto/pull/7982/commits/77ec5e9c5cacc7d41811be6f133b1b117d21fd7b#diff-838efde24d15b34a148c749a8e274206R39 would be a Lookup impl that you could safely re-use here when Lookup not provided.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java (hunk 9 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java]: How is that possible if I didn't change the behavior here?
The `extractUnique(PlanNode)` is delegating to the `extractUnique(PlanNode, Lookup)` which is almost identical to the previous implementation of `extractUnique` (which underneath used `extractExpressions` which before and now are recursive)?

Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java]: You can still use `ExpressionExtractor.extractExpressionsNonRecursive` if needed but there was not equivalent `extractUnique` before nor it exists now.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionExtractor.java (hunk 8 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionExtractor.java]: `extractExpressions` without `Lookup` given should not be used for iterative plans.

I rethought that though and I'm falling back to what I introduced already - relying on `Optional<Lookup>` in `ExpressionExtractor.Visitor`. This does not require any special handling in `Lookup` despite what's there already.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java (hunk 10 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java]: I think you can extract it from here in your PR (or if your PR lands earlier, I'll re-use it here).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 11 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: It's smaller scope, so we don't pollute this class with more details than it needs.
We've done such a scope truncation before. Mind the `TransformExistsApplyToScalarApply` or `PartialAggregationPushDown`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java (hunk 12 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: Yours would fail here too.
I think it's better to be explicit about optionality of `Lookup` and eventually remove that bit (`Optional` and non-lookup methods) from `Visitor`, `DependencyExtractor` and `ExpressionsExtractor`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java (hunk 9 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java]: > How is that possible if I didn't change the behavior here?

What has been changed is that you added handling of `GroupReferences`: "if instanceof GroupReference, lookup & recurse"

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java (hunk 12 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: Of course. It would explicitly reject `GroupReference` instead of `StackoverflowError` :)

**Comment by maciejgrzybek** [comment]
The product test failure is existing on master already, it was no introduced by this PR.
When master becomes green, I'll rebase.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java (hunk 12 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: Have a look how I've done this currently.
https://github.com/prestodb/presto/pull/7429/commits/0d628d8a47a35682963568ec68ded5a5bbecb80e#diff-fb5dfcc32facc6f3778ebc715736dd4eR40 + https://github.com/prestodb/presto/pull/7429/commits/0d628d8a47a35682963568ec68ded5a5bbecb80e#diff-fb5dfcc32facc6f3778ebc715736dd4eR70

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java (hunk 9 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DependencyExtractor.java]: `DependencyExtractor.extractUnique(PlanNode node)` has not changed. It's underlying `ExpressionExtractor.extractExpressions(PlanNode plan)` neither. `ExpressionExtractor.Visitor` which is used in `extractExpressions` by default takes `Optional.empty()` in c-tor and the only change to the visitor is in `visitPlan` which only in case there's a `GroupReference` node visited, it resolves it using given `Lookup` and tries to visit the appropriate node (instead of `GroupReference` mock).
In other words:
1. If you don't use iterative plans, nothing has changed for you
2. If you use iterative plans, the only difference is you have to provide a `Lookup` and when there's a `GroupReference`, it's resolved to corresponding node, although there's no extra recursion added beyond what was there before (e.g. aforementioned `ExpressionExtractor.extractExpressionsNonRecursive`.

The default `extractUnique` didn't change the behavior.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java (hunk 12 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: So if I call with `recursive=false`, children (whether `GroupReferences`  or not) won't be inspected? Looks good!

Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java]: Yup, that's correct.
`visitPlan` is called only when there's no appropriate `visitX` for X node.
Then, I check if `GroupReference` was seen (it happens in `visitPlan`), if yes, then assuming `Lookup` was provided (asserted) it's used to resolve such node and use appropriate `visitX` for it. No recursion, unless requested (`recursive` flag set by `ExpressionExtractor.extractExpressionsNonRecursive` method).

**Comment by maciejgrzybek** [comment]
Seems it's ready for you, @martint.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java [hunk 13]
```diff
@@ -43,6 +43,7 @@
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Iterables;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java]: I'd change the title of this commit to "Preserve output shape in ScalarAggregationToJoinRewriter"

**Comment by maciejgrzybek** [comment]
@martint: rephrased `Limit output symbols [...]` into `Preserve output shape [...]`, as requested. Also changed (in the commit 881fcff) the `if-else` on `Optional` into functional-style expression in both `ScalarAggregationToJoinRewriter.resolve(PlanNode)` and `ScalarAggregationToJoinRewriter.searchFrom(PlanNode)` for consistency.
